pkg/agent/core/dumpstate.go

```diff
@@ -3,142 +3,83 @@ package core
-			At:        s.lastRequest.at,
-			Resources: s.lastRequest.resources,
-		}
-	}
-
-	return pluginStateDump{
-		OngoingRequest: s.ongoingRequest,
-		ComputeUnit:    shallowCopy(s.computeUnit),
-		LastRequest:    lastRequest,
-		LastFailureAt:  shallowCopy(s.lastFailureAt),
-		Permit:         shallowCopy(s.permit),
+func (s *pluginState) deepCopy() pluginState {
+	return pluginState{
+		OngoingRequest: s.OngoingRequest,
+		ComputeUnit:    shallowCopy[api.Resources](s.ComputeUnit),
+		LastRequest: shallowCopy(s.LastRequest, func(r pluginRequested) pluginRequested {
+			return pluginRequested{At: r.At, Resources: r.Resources}
+		}),
+		LastFailureAt: shallowCopy[time.Time](s.LastFailureAt),
+		Permit:        shallowCopy[api.Resources](s.Permit),
 	}
 }
 
-type monitorStateDump struct {
-	OngoingRequest     *OngoingMonitorRequestDump `json:"ongoingRequest"`
-	RequestedUpscale   *requestedUpscaleDump      `json:"requestedUpscale"`
-	DeniedDownscale    *deniedDownscaleDump       `json:"deniedDownscale"`
-	Approved           *api.Resources             `json:"approved"`
-	DownscaleFailureAt *time.Time                 `json:"downscaleFailureAt"`
-	UpscaleFailureAt   *time.Time                 `json:"upscaleFailureAt"`
-}
-type OngoingMonitorRequestDump struct {
-	Kind      monitorRequestKind `json:"kind"`
-	Requested api.Resources      `json:"resources"`
-}
-type requestedUpscaleDump struct {
-	At        time.Time         `json:"at"`
-	Base      api.Resources     `json:"base"`
-	Requested api.MoreResources `json:"requested"`
-}
-type deniedDownscaleDump struct {
-	At        time.Time     `json:"at"`
-	Current   api.Resources `json:"current"`
-	Requested api.Resources `json:"requested"`
-}
-
-func (s *monitorState) dump() monitorStateDump {
-	var requestedUpscale *requestedUpscaleDump
-	if s.requestedUpscale != nil {
-		requestedUpscale = &requestedUpscaleDump{
-			At:        s.requestedUpscale.at,
-			Base:      s.requestedUpscale.base,
-			Requested: s.requestedUpscale.requested,
-		}
-	}
-
-	var deniedDownscale *deniedDownscaleDump
-	if s.deniedDownscale != nil {
-		deniedDownscale = &deniedDownscaleDump{
-			At:        s.deniedDownscale.at,
-			Current:   s.deniedDownscale.current,
-			Requested: s.deniedDownscale.requested,
-		}
-	}
-
-	var ongoingRequest *OngoingMonitorRequestDump
-	if s.ongoingRequest != nil {
-		ongoingRequest = &OngoingMonitorRequestDump{
-			Kind:      s.ongoingRequest.kind,
-			Requested: s.ongoingRequest.requested,
-		}
-	}
-
-	return monitorStateDump{
-		OngoingRequest:     ongoingRequest,
-		RequestedUpscale:   requestedUpscale,
-		DeniedDownscale:    deniedDownscale,
-		Approved:           shallowCopy(s.approved),
-		DownscaleFailureAt: shallowCopy(s.downscaleFailureAt),
-		UpscaleFailureAt:   shallowCopy(s.upscaleFailureAt),
+func (s *monitorState) deepCopy() monitorState {
+	return monitorState{
+		OngoingRequest: shallowCopy(s.OngoingRequest, func(r ongoingMonitorRequest) ongoingMonitorRequest {
```

Comment:
I don't understand how this function `func(r ongoingMonitorRequest) ongoingMonitorRequest` helps here. If this is going to be a shallow copy then copying the object value and making a pointer to the new value should be enough.

Reply:
Yeah, that's fair. There's a number of those. I was _kind of_ thinking along the lines of "how do we make sure this is still correct if we add a field that needs deeper copying?".
But maybe that's more hassle than it's worth, and it's better to just be simple here.

Reply:
55dc0b786b558aac16ebf8c200a877f948c31ece - overwritten by later rebase.

Reply:
Can we remove the extra argument (`f ...func(T) T`) from `shallowCopy`? We haven't used it anywhere and it's confusing.

Reply:
c4b5dc5e1fe28e912ce6aa2312e66e317aa678aa - same deal w/ rebase.